handler/rfc8628/strategy_hmacsha.go

```diff
@@ -0,0 +1,79 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/ory/x/errorsx"
+	"github.com/ory/x/randx"
+
+	"github.com/ory/fosite"
+	enigma "github.com/ory/fosite/token/hmac"
+)
+
+type DefaultDeviceStrategy struct {
+	RFC8628CodeStrategy
+
+	Enigma *enigma.HMACStrategy
+	Config interface {
+		fosite.DeviceAndUserCodeLifespanProvider
+	}
+}
+
+func (h *DefaultDeviceStrategy) GenerateUserCode(ctx context.Context) (token string, signature string, err error) {
+	seq, err := randx.RuneSequence(8, []rune("BCDFGHJKLMNPQRSTVWXZ"))
+	if err != nil {
+		return "", "", err
+	}
+	userCode := string(seq)
```

Comment:
I think that's quite implementation specific, iv seen providers skip dashes ect

I dont this the specs have any recommendations for the formatting, just an example with a dash.

However the length of this probably should be configurable (with restrictions)

Reply:
Agreed about configurable length :)

I would not set a dash as it require apps implementing this DAG feature to show it in their UI to display it properly.
Also, it can be quite tricky as people could not understand that they need to you the code with the dash.

Reply:
> Also, it can be quite tricky as people could not understand that they need to you the code with the dash.

I'd strip the dashes in the response, so that codes with and without dashes work the same.


> I would not set a dash as it require apps implementing this DAG feature to show it in their UI to display it properly.

Simple UIs that just show the code verbatim and present the user with a text box will have improved UX because the user can more easily scan and enter the code.

More sophisticated UIs can strip the dashes themselves, have a single input field per letter that automatically jumps to the next one, etc.